### ISSUE TITLE
[Proposal] Optional environment specific configurations

### ISSUE THREAD

**Issue by stefanbauer** [laravel/ideas#1110]
Had a talk now with different people about the following situation and most of them have the same problems.

Imagine you have a complicated logging stack. Different channels, different levels. Thanks to the new logging in L5.6 it's really easy to configure. But the main problem is, that logging is different in local, staging or production environment. And if you have a lot of channels for different purposes, you know what I'm talking about. Anyways. The log levels are very different, because you will in most cases never have a level debug on production machines, may be just some criticals for slack. But locally or on staging, the level are different for different channels. 

To achieve that, there are not much possibilities. Either you use the `.env` file for that or you create some hacky stuff in the configuration files with some conditions like `if env('APP_ENV') === 'production'` then override this configuration with that. Both cases are bad decision imho. `.env` is not the right place for that stuff, because this kind of configuration is **static**. So it's already known before you ever deploy. This is also true for some other values. Like `APP_DEBUG`. I know, I would never set this to true on production. But maybe in staging.

I know, introducing the `.env` stuff makes Laravel simple. And it's cool. But maybe it could be an idea, to introduce optional environment specific configuration for people, who would like to use it (for the use cases above). Some would say: "Hey, use your .env". But to be honest... If you have 10 production servers, and you would like to reconfigure a little bit the production logging, you have to log in into 10 servers and edit the `.env`? Really? My proposal therefore is, (just an example) to load (optionally and if existing and if on this environment) something like `config/queues.production.php`. This will be merged with `config/queues.php` with precedence. With that, you can specify (you **can**, you don't have to) different environment configurations at the same time. The cool side effect is, that the `.env` can be very small and simple, because it's just necessary to store sensitive data there.

But again. Developers can use it, but don't have to. They could just work as they did.

Before diving into it and creating a PR, I would just like to ask, if this is something, that could be interesting?

**Comment by tomschlick** [comment]
> If you have 10 production servers, and you would like to reconfigure a little bit the production logging, you have to log in into 10 servers and edit the .env? Really?

That's what configuration management tools are for. If you want something simpler than that, [Envoyer](https://envoyer.io/) handles .env files across multiple servers.

**Comment by stefanbauer** [comment]
Yeah I get your idea, but as `.env` is just a simple key/value file, I don't see any benefit here. If you like to define some logging channels for instance on production only, this is absolutely uncomfortable with an `.env`. You don't agree that it would be much easier to just define that channel in a `logging.production.php` and no elsewhere and work is done? Otherwise it's really hard to deal with it, I think.

**Comment by stefanbauer** [comment]
I mean: There is already something like this. Basically at least. If you define your `APP_ENV` for instance on your php-fpm like `production`, a environment specific env-file will be loaded (if existing). So if a `.env.production` exists, this will be loaded. So there is already kind of that for the situations I describe.

https://github.com/laravel/framework/blob/5.6/src/Illuminate/Foundation/Bootstrap/LoadEnvironmentVariables.php#L57

**Comment by martinbean** [comment]
@stefanbauer Laravel _used_ to have environment-specific configuration files like you describe, and there will have a been a reason for moving away from them. Its environment should determine how your application should run, and *.env* files merely adopt the role of environment variables in situations where they can’t be used (i.e. serving locally).

If you absolutely need environment-specific configurations for say, queues, then there’s nothing stopping you adding an entry to the **config/queues.php** file:

```php
return [
    'connections' => [
        'production' => [
            'driver' => 'sqs',
            // Other production-specific values
        ],
    ],
];
```
```
APP_ENV=production
QUEUE_DRIVER=production
```
But I really can’t see what would be so special about production environment that it warrants merging in configurations from multiple files?

As for:

> If you have 10 production servers, and you would like to reconfigure a little bit the production logging, you have to log in into 10 servers and edit the .env? Really?

If I’m managing 10 servers then I wouldn’t be manually logging in, and I definitely wouldn’t be editing production configuration files by hand. You’d use some sort of service to make updating your servers’s configuration automated.

**Comment by martinbean** [comment]
@stefanbauer I feel as though you’re going about configuration the wrong way. You should be configuring your stacks etc in your configuration files, and then using environment variables to specify the configuration you want to use. As you’ve found: environment variables are for just key-value pairs, not defining your application’s configuration.

If you _do_ have such environment-specific configurations then create stacks for them in your config files and then choose that stack in your **.env** file.

**Comment by stefanbauer** [comment]
Thanks for your reply man. We get closer step by step. You're totally right. But this is just right, if you have only 1 stack, then, yeah, you can set one key/value in the `.env`. But if you have more channels, more stacks, for different purposes, on different, places, it's not enough to set just one value in the `.env`. You know what i mean?

**Comment by martinbean** [comment]
@stefanbauer I think you’re missing what I’m trying to get at, though. If you need to say, log to multiple stacks, then you can “compose” them in a new logging configuration:

```php
return [
    'channels' => [
        'production' => [
            'driver' => 'stack',
            'channels' => [
                // Array of channels you need to log to in production
            ],
        ],
    ],
];
```
```
APP_ENV=production
LOG_CHANNEL=production
```

Am I misunderstanding you?

**Comment by stefanbauer** [comment]
Nah you get me right I think. And you totally agree what you describe, if you have only 1 stack. If you have several separated channels, several stacks (nobody forces you to have only 1 stack), then it's not only setting one key/value.

We have another application that as at least 10 channels. Event, Router, Legacy, Security, and for several APIs. You have different configurations for all the channels. Different log levels, and different settings. And additional to that you have several stacks, that are differently configured for different environments, because in production, some channels are not used, because only used for debugging, for instance, or in production just log from `alert`, instead of `error`, or so. Got it?

**Comment by AndreSchwarzer** [comment]
Hey @stefanbauer have you found a clean implementation in the mean time?
